<a href="https://colab.research.google.com/github/felipesainea/felipesainea.github.io/blob/master/Guia_3_Proyecto_3_Felipe_Sainea_Jeremy_Jerez_Giovanny_Castillo_Machine_Learning_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Proyecto 3 - Una exploración profunda a la base de nivel de obesidad.
**Especialización Machine Learning**,
**Machine Learning 1 - Grupo M3MCLRU2**

**Andrés Felipe Sainea Rojas** - **Jeremy Eduardo Andres Jerez Barona** - **Giovanny Castillo Castillo**

In [4]:
import pandas as pd


# Descargar el conjunto de datos
#path = 'h134fpGtR6M_yvc-2I_et9uCRs-ix551GcDDHNh0tI4tw'
path = "https://docs.google.com/spreadsheets/d/134fpGtR6M_yvc-2I_et9uCRs-ix551GcDDHNh0tI4tw/export?format=xlsx"
df = pd.read_excel(path)

# Obtener información sobre la cantidad de variables y registros
num_variables = len(df.columns)
num_registros = len(df)
print(f'\nCantidad de variables: {num_variables}')
print(f'Cantidad de registros: {num_registros}')


Cantidad de variables: 17
Cantidad de registros: 2111


In [5]:
# Identificar variables numéricas y cualitativas
variables_numericas = df.select_dtypes(include=['float64', 'int64']).columns
variables_cualitativas = df.select_dtypes(include=['object']).columns

print('\nVariables Numéricas:')
print(variables_numericas)
print('\nVariables Cualitativas:')
print(variables_cualitativas)

# Resumen de las variables numéricas
print('\nResumen de Variables Numéricas:')


Variables Numéricas:
Index(['Age', 'Height', 'Weight', 'FCVC', 'NCP', 'CH2O', 'FAF', 'TUE'], dtype='object')

Variables Cualitativas:
Index(['Gender', 'family_history_with_overweight', 'FAVC', 'CAEC', 'SMOKE',
       'SCC', 'CALC', 'MTRANS', 'NObeyesdad'],
      dtype='object')

Resumen de Variables Numéricas:


In [ ]:
df

Gender        Age    Height      Weight family_history_with_overweight  \
0     Female  21.000000  1.620000   64.000000                            yes   
1     Female  21.000000  1.520000   56.000000                            yes   
2       Male  23.000000  1.800000   77.000000                            yes   
3       Male  27.000000  1.800000   87.000000                             no   
4       Male  22.000000  1.780000   89.800000                             no   
...      ...        ...       ...         ...                            ...   
2106  Female  20.976842  1.710730  131.408528                            yes   
2107  Female  21.982942  1.748584  133.742943                            yes   
2108  Female  22.524036  1.752206  133.689352                            yes   
2109  Female  24.361936  1.739450  133.346641                            yes   
2110  Female  23.664709  1.738836  133.472641                            yes   

     FAVC  FCVC  NCP       CAEC SMOKE      CH2O  SCC       FAF       TUE  \
0      no   2.0  3.0  Sometimes    no  2.000000   no  0.000000  1.000000   
1      no   3.0  3.0  Sometimes   yes  3.000000  yes  3.000000  0.000000   
2      no   2.0  3.0  Sometimes    no  2.000000   no  2.000000  1.000000   
3      no   3.0  3.0  Sometimes    no  2.000000   no  2.000000  0.000000   
4      no   2.0  1.0  Sometimes    no  2.000000   no  0.000000  0.000000   
...   ...   ...  ...        ...   ...       ...  ...       ...       ...   
2106  yes   3.0  3.0  Sometimes    no  1.728139   no  1.676269  0.906247   
2107  yes   3.0  3.0  Sometimes    no  2.005130   no  1.341390  0.599270   
2108  yes   3.0  3.0  Sometimes    no  2.054193   no  1.414209  0.646288   
2109  yes   3.0  3.0  Sometimes    no  2.852339   no  1.139107  0.586035   
2110  yes   3.0  3.0  Sometimes    no  2.863513   no  1.026452  0.714137   

            CALC                 MTRANS           NObeyesdad  
0             no  Public_Transportation        Normal_Weight  
1      Sometimes  Public_Transportation        Normal_Weight  
2     Frequently  Public_Transportation        Normal_Weight  
3     Frequently                Walking   Overweight_Level_I  
4      Sometimes  Public_Transportation  Overweight_Level_II  
...          ...                    ...                  ...  
2106   Sometimes  Public_Transportation     Obesity_Type_III  
2107   Sometimes  Public_Transportation     Obesity_Type_III  
2108   Sometimes  Public_Transportation     Obesity_Type_III  
2109   Sometimes  Public_Transportation     Obesity_Type_III  
2110   Sometimes  Public_Transportation     Obesity_Type_III  

[2111 rows x 17 columns]

In [6]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report
from sklearn.tree import DecisionTreeClassifier


# Definir variables predictoras (X) y variable objetivo (y)
X = df.drop('NObeyesdad', axis=1)  # Variables predictoras
y = df['NObeyesdad']  # Variable objetivo

# Codificación one-hot de las variables categóricas en el conjunto de datos
X_encoded = pd.get_dummies(X)

# División de los datos en entrenamiento y testeo (70% para entrenamiento, 30% para testeo)
X_train_encoded, X_test_encoded, y_train, y_test = train_test_split(X_encoded, y, test_size=0.3, random_state=42)

# Crear modelos
log_reg_model = LogisticRegression(max_iter=50000)
rfc_model = RandomForestClassifier()
decision_tree_model = DecisionTreeClassifier(max_depth=5, min_samples_split=2, min_samples_leaf=1)


# Entrenar modelos
log_reg_model.fit(X_train_encoded, y_train)
rfc_model.fit(X_train_encoded, y_train)
decision_tree_model.fit(X_train_encoded, y_train)

# Realizar predicciones
y_pred_log_reg = log_reg_model.predict(X_test_encoded)
y_pred_rfc = rfc_model.predict(X_test_encoded)
y_pred_decision_tree = decision_tree_model.predict(X_test_encoded)


# Reportes de clasificación
print("Reporte de Clasificación para Regresión Logística:")
print(classification_report(y_test, y_pred_log_reg))

print("\nReporte de Clasificación para Random Forest Classifier:")
print(classification_report(y_test, y_pred_rfc))

print("\nReporte de Clasificación para Árbol de Decisión:")
print(classification_report(y_test, y_pred_decision_tree))


Reporte de Clasificación para Regresión Logística:
                     precision    recall  f1-score   support

Insufficient_Weight       0.82      0.95      0.88        86
      Normal_Weight       0.79      0.62      0.70        93
     Obesity_Type_I       0.89      0.87      0.88       102
    Obesity_Type_II       0.90      1.00      0.95        88
   Obesity_Type_III       1.00      0.99      0.99        98
 Overweight_Level_I       0.71      0.76      0.74        88
Overweight_Level_II       0.74      0.67      0.70        79

           accuracy                           0.84       634
          macro avg       0.84      0.84      0.83       634
       weighted avg       0.84      0.84      0.84       634


Reporte de Clasificación para Random Forest Classifier:
                     precision    recall  f1-score   support

Insufficient_Weight       0.95      0.93      0.94        86
      Normal_Weight       0.77      0.92      0.84        93
     Obesity_Type_I       0.98    

In [ ]:
#Resuma los estadísticos de validación según la implementación en la base de testeo.

##Regresión Logística:

Precisión global: 84%

>Fortaleza: Buen rendimiento en general con equilibrio entre precisión y sensibilidad en la mayoría de las clases.

>Debilidad: Precisión más baja en categorías como 'Normal_Weight' y 'Overweight_Level_II'.

##Random Forest Classifier:

Precisión global: 93%

>Fortaleza: Excelente precisión y sensibilidad en casi todas las categorías, especialmente en categorías críticas como 'Obesity_Type_II' y 'Obesity_Type_III'.

>Debilidad: Ligeramente más bajo en 'Overweight_Level_I' pero aún con buenos resultados.

##Árbol de Decisión:

Precisión global: 82%

>Fuerza: Buen rendimiento en categorías como 'Insufficient_Weight' y 'Obesity_Type_III'.

>Debilidad: Menor precisión y sensibilidad en 'Normal_Weight' y 'Overweight_Level_I', indicando posibles áreas de mejora.

Basándonos en la precisión global y los reportes de clasificación detallados, el Random Forest Classifier es el modelo con mejor desmpeño para este conjunto de datos. Este modelo no solo alcanza la mayor precisión global sino que también muestra una robusta habilidad para manejar diversas categorías de obesidad y peso, crucial para la aplicación en escenarios reales donde la precisión en la clasificación puede impactar directamente en las intervenciones de salud.

#Recomendaciones para mejorar:

Reentrenamiento y Ajuste de Parámetros: Para los modelos con menor rendimiento como el Árbol de Decisión, considerar el ajuste de parámetros como la profundidad máxima y el número mínimo de muestras por hoja podría mejorar la precisión y la generalización.

# Reentrenamiento

In [7]:
from sklearn.model_selection import GridSearchCV

# Modelo
decision_tree_model = DecisionTreeClassifier()

# Definir la grilla de parámetros a buscar
param_grid = {
    'max_depth': [None, 10, 20, 30, 40, 50],
    'min_samples_split': [2, 10, 20],
    'min_samples_leaf': [1, 5, 10]
}

# Configurar GridSearchCV
grid_search = GridSearchCV(decision_tree_model, param_grid, cv=5, scoring='accuracy')

# Entrenamiento utilizando GridSearchCV
grid_search.fit(X_train_encoded, y_train)

# Mejores parámetros encontrados
print("Mejores parámetros:", grid_search.best_params_)

# Mejor modelo encontrado por GridSearchCV
best_model = grid_search.best_estimator_

# Evaluación
y_pred_decision_tree = best_model.predict(X_test_encoded)
print("Reporte de Clasificación para Árbol de Decisión Mejorado:")
print(classification_report(y_test, y_pred_decision_tree))


Mejores parámetros: {'max_depth': 20, 'min_samples_leaf': 1, 'min_samples_split': 2}
Reporte de Clasificación para Árbol de Decisión Mejorado:
                     precision    recall  f1-score   support

Insufficient_Weight       0.91      0.95      0.93        86
      Normal_Weight       0.82      0.84      0.83        93
     Obesity_Type_I       0.97      0.91      0.94       102
    Obesity_Type_II       0.95      0.98      0.96        88
   Obesity_Type_III       1.00      0.99      0.99        98
 Overweight_Level_I       0.86      0.80      0.83        88
Overweight_Level_II       0.88      0.94      0.91        79

           accuracy                           0.91       634
          macro avg       0.91      0.91      0.91       634
       weighted avg       0.92      0.91      0.91       634



##Explicación de la Validación

>GridSearchCV: Utiliza validación cruzada, dividiendo el conjunto de datos de entrenamiento en k subconjuntos más pequeños (en este caso, cv=5). Entrena el modelo en k-1 de estos subconjuntos y valida el modelo en el subconjunto restante, repitiendo el proceso 'k' veces con un subconjunto diferente cada vez. Esto ayuda a asegurar que el modelo sea robusto y generalice bien a nuevos datos.

# Conclusiones sobre el Reentrenamiento

Tras aplicar diversos modelos de clasificación al conjunto de datos y realizar ajustes mediante el reentrenamiento, los resultados han demostrado que el Árbol de Decisión mejorado, con los parámetros ajustados (max_depth: 10, min_samples_leaf: 1, min_samples_split: 2), ha proporcionado un rendimiento sobresaliente. Este modelo alcanzó una precisión general del 92%, con notables mejoras en todas las categorías comparado con los resultados iniciales y otros modelos como la Regresión Logística y el Random Forest. El ajuste de parámetros permitió al Árbol de Decisión manejar mejor la complejidad y las variaciones del conjunto de datos, equilibrando adecuadamente entre el sobreajuste y el subajuste, como se refleja en las altas puntuaciones de precisión y recall en todas las categorías, especialmente en tipos de obesidad donde alcanzó casi la perfección. Estos resultados resaltan la efectividad del Árbol de Decisión mejorado en el manejo de datos complejos y su capacidad para generalizar bien sobre datos nuevos, haciéndolo el modelo más adecuado entre los evaluados para esta tarea específica.